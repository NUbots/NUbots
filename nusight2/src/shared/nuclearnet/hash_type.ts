import XXH from "xxhashjs";

// Key for hashing NUClear types and subtypes.
const hashKey = 0x4e55436c;

/** Hash the given message type using the same algorithm used in NUClear */
export function hashType(type: string): Buffer {
  // Matches hashing implementation from NUClearNet
  // See https://github.com/Fastcode/NUClearNet.js/blob/857ed674/src/NetworkBinding.cpp#L32
  let hashString: string = XXH.h64(type, hashKey).toString(16);
  // The hash string may truncate if it's smaller than 16 characters so we pad it with 0s
  hashString = ("0".repeat(16) + hashString).slice(-16);

  return Buffer.from((hashString.match(/../g) as string[]).reverse().join(""), "hex");
}

/** Hash the given message subtype using the same algorithm used in NUClear */
export function hashSubtype(type: string): number {
  return XXH.h32(type, hashKey).toNumber();
}
