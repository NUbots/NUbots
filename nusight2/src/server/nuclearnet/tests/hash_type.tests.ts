import { describe, expect, it } from "vitest";
import { describe, expect, it } from "vitest";
import { hashType } from "../hash_type";

describe("hashType()", () => {
  it("matches previous snapshots", () => {
    expect(hashType("foo")).toMatchSnapshot();
    expect(hashType("bar")).toMatchSnapshot();
    expect(hashType("baz")).toMatchSnapshot();
    expect(hashType("cheesecake")).toMatchSnapshot();
  });

  it("ensures it returns 8 bytes", () => {
    expect(hashType("foo").byteLength).toBe(8);
    expect(hashType("bar").byteLength).toBe(8);
    expect(hashType("baz").byteLength).toBe(8);
    // "cheesecake" happens to hash to 7 bytes without padding.
    expect(hashType("cheesecake").byteLength).toBe(8);
  });
});
