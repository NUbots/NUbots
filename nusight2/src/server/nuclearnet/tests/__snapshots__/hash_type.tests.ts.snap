// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hashType() > matches previous snapshots 1`] = `
{
  "data": [
    225,
    233,
    240,
    189,
    213,
    194,
    172,
    79,
  ],
  "type": "Buffer",
}
`;

exports[`hashType() > matches previous snapshots 2`] = `
{
  "data": [
    188,
    214,
    134,
    145,
    61,
    19,
    178,
    202,
  ],
  "type": "Buffer",
}
`;

exports[`hashType() > matches previous snapshots 3`] = `
{
  "data": [
    11,
    128,
    31,
    158,
    81,
    99,
    174,
    59,
  ],
  "type": "Buffer",
}
`;

exports[`hashType() > matches previous snapshots 4`] = `
{
  "data": [
    89,
    255,
    223,
    146,
    90,
    203,
    129,
    10,
  ],
  "type": "Buffer",
}
`;

exports[`hashType() matches previous snapshots 1`] = `
{
  "data": [
    225,
    233,
    240,
    189,
    213,
    194,
    172,
    79,
  ],
  "type": "Buffer",
}
`;

exports[`hashType() matches previous snapshots 2`] = `
{
  "data": [
    188,
    214,
    134,
    145,
    61,
    19,
    178,
    202,
  ],
  "type": "Buffer",
}
`;

exports[`hashType() matches previous snapshots 3`] = `
{
  "data": [
    11,
    128,
    31,
    158,
    81,
    99,
    174,
    59,
  ],
  "type": "Buffer",
}
`;

exports[`hashType() matches previous snapshots 4`] = `
{
  "data": [
    89,
    255,
    223,
    146,
    90,
    203,
    129,
    10,
  ],
  "type": "Buffer",
}
`;
